src/controllers/colormapper.cpp

```diff
@@ -0,0 +1,59 @@
+#include "controllers/colormapper.h"
+
+#include <cmath>
+#include <cstdint>
+
+#include "util/debug.h"
+
+namespace {
+double colorDistance(QRgb a, QRgb b) {
+    // This algorithm calculates the distance between two colors. In
+    // contrast to the L2 norm, this also tries take the human perception
+    // of colors into account. More accurate algorithms like the CIELAB2000
+    // Delta-E rely on sophisticated color space conversions and need a lot
+    // of costly computations. In contrast, this is a low-cost
+    // approximation and should be sufficently accurate.
+    // More details: https://www.compuphase.com/cmetric.htm
+    long mean_red = (static_cast<long>(qRed(a)) + static_cast<long>(qRed(b))) / 2;
+    long delta_red = static_cast<long>(qRed(a)) - static_cast<long>(qRed(b));
+    long delta_green = static_cast<long>(qGreen(a)) - static_cast<long>(qGreen(b));
+    long delta_blue = static_cast<long>(qBlue(a)) - static_cast<long>(qBlue(b));
+    return sqrt(
+            (((512 + mean_red) * delta_red * delta_red) >> 8) +
+            (4 * delta_green * delta_green) +
+            (((767 - mean_red) * delta_blue * delta_blue) >> 8));
+}
+} // namespace
+
+QPair<QRgb, QVariant> ColorMapper::getNearestColor(QRgb desiredColor) {
+    // If desired color is already in cache, use cache entry
+    QMap<QRgb, QRgb>::const_iterator i = m_cache.find(desiredColor);
+    QMap<QRgb, QVariant>::const_iterator j;
+    if (i != m_cache.constEnd()) {
+        j = m_availableColors.find(i.value());
+        DEBUG_ASSERT(j != m_availableColors.constEnd());
+        qDebug() << "ColorMapper cache hit for" << desiredColor << ":"
```

Comment:
how often is this debug going to be printed, and do we need it?  I've started putting debugs behind if statements to reduce noise in our logs, like:

constexpr bool COLORMAP_DEBUG = false;

...

if (COLORMAP_DEBUG) qDebug() << "blah";

Reply:
Our simple Logger class allows distinguishes between *trace* and *debug* levels. Use *trace*, because *debug* is always written.

Cluttering the code with individual constants in every single source file is not recommended.